File: services/k8s-service-go/internal/routes/configuration.go
package routes

import (
	"github.com/go-chi/chi/v5"

	"github.com/junginho0901/kubeast/services/k8s-service-go/internal/handler"
)

// RegisterConfiguration — ConfigMap / Secret / generic resources +
// scheduling-and-policy resources (HPA, VPA, PDB, PriorityClass,
// RuntimeClass, ResourceQuota, LimitRange) + admission-webhook configs
// + Lease.
//
// Grouping rationale: these are all "configuration & policy" objects
// that sit one layer below workload spec. Splitting hairs (HPA into
// workloads.go, etc.) would scatter related routes across files.
func RegisterConfiguration(r chi.Router, h *handler.Handler) {
	// ConfigMaps
	r.Get("/api/v1/configmaps/all", h.GetAllConfigMaps)
	r.Get("/api/v1/namespaces/{namespace}/configmaps", h.GetConfigMaps)
	r.Get("/api/v1/namespaces/{namespace}/configmaps/{name}/describe", h.DescribeConfigMap)
	r.Get("/api/v1/namespaces/{namespace}/configmaps/{name}/yaml", h.GetConfigMapYAML)
	r.Delete("/api/v1/namespaces/{namespace}/configmaps/{name}", h.DeleteConfigMap)

	// Secrets
	r.Get("/api/v1/secrets/all", h.GetAllSecrets)
	r.Get("/api/v1/namespaces/{namespace}/secrets", h.GetSecrets)
	r.Get("/api/v1/namespaces/{namespace}/secrets/{name}/describe", h.DescribeSecret)
	r.Get("/api/v1/namespaces/{namespace}/secrets/{name}/yaml", h.GetSecretYAML)
	r.Delete("/api/v1/namespaces/{namespace}/secrets/{name}", h.DeleteSecret)

	// Generic resources
	r.Get("/api/v1/resources", h.GetGenericResources)
	r.Post("/api/v1/search", h.SearchResources)
	r.Get("/api/v1/resources/json", h.GetGenericResourceJSON)
	r.Get("/api/v1/resources/yaml", h.GetGenericResourceYAML)
	r.Post("/api/v1/resources/yaml/apply", h.ApplyResourceYAML)
	r.Post("/api/v1/resources/yaml/create", h.CreateResourcesFromYAML)
	r.Get("/api/v1/resources/describe", h.DescribeGenericResource)

	// HPA
	r.Get("/api/v1/hpas/all", h.GetAllHPAs)
	r.Get("/api/v1/namespaces/{namespace}/hpas", h.GetHPAs)
	r.Get("/api/v1/namespaces/{namespace}/hpas/{name}/describe", h.DescribeHPA)
	r.Get("/api/v1/namespaces/{namespace}/hpas/{name}/yaml", h.GetHPAYAML)
	r.Delete("/api/v1/namespaces/{namespace}/hpas/{name}", h.DeleteHPA)

	// VPA
	r.Get("/api/v1/vpas/all", h.GetAllVPAs)
	r.Get("/api/v1/namespaces/{namespace}/vpas", h.GetVPAs)
	r.Get("/api/v1/namespaces/{namespace}/vpas/{name}/describe", h.DescribeVPA)
	r.Get("/api/v1/namespaces/{namespace}/vpas/{name}/yaml", h.GetVPAYAML)
	r.Delete("/api/v1/namespaces/{namespace}/vpas/{name}", h.DeleteVPA)

	// PDB
	r.Get("/api/v1/pdbs/all", h.GetAllPDBs)
	r.Get("/api/v1/namespaces/{namespace}/pdbs", h.GetPDBs)
	r.Get("/api/v1/namespaces/{namespace}/pdbs/{name}/describe", h.DescribePDB)
	r.Get("/api/v1/namespaces/{namespace}/pdbs/{name}/yaml", h.GetPDBYAML)
	r.Delete("/api/v1/namespaces/{namespace}/pdbs/{name}", h.DeletePDB)

	// PriorityClass (cluster-scoped)
	r.Get("/api/v1/priorityclasses", h.GetPriorityClasses)
	r.Get("/api/v1/priorityclasses/{name}/describe", h.DescribePriorityClass)
	r.Get("/api/v1/priorityclasses/{name}/yaml", h.GetPriorityClassYAML)
	r.Delete("/api/v1/priorityclasses/{name}", h.DeletePriorityClass)

	// RuntimeClass (cluster-scoped)
	r.Get("/api/v1/runtimeclasses", h.GetRuntimeClasses)
	r.Get("/api/v1/runtimeclasses/{name}/describe", h.DescribeRuntimeClass)
	r.Get("/api/v1/runtimeclasses/{name}/yaml", h.GetRuntimeClassYAML)
	r.Delete("/api/v1/runtimeclasses/{name}", h.DeleteRuntimeClass)

	// ResourceQuota (namespace-scoped)
	r.Get("/api/v1/resourcequotas/all", h.GetAllResourceQuotas)
	r.Get("/api/v1/namespaces/{namespace}/resourcequotas", h.GetResourceQuotas)
	r.Get("/api/v1/namespaces/{namespace}/resourcequotas/{name}/describe", h.DescribeResourceQuota)
	r.Get("/api/v1/namespaces/{namespace}/resourcequotas/{name}/yaml", h.GetResourceQuotaYAML)
	r.Delete("/api/v1/namespaces/{namespace}/resourcequotas/{name}", h.DeleteResourceQuota)

	// LimitRange (namespace-scoped)
	r.Get("/api/v1/limitranges/all", h.GetAllLimitRanges)
	r.Get("/api/v1/namespaces/{namespace}/limitranges", h.GetLimitRanges)
	r.Get("/api/v1/namespaces/{namespace}/limitranges/{name}/describe", h.DescribeLimitRange)
	r.Get("/api/v1/namespaces/{namespace}/limitranges/{name}/yaml", h.GetLimitRangeYAML)
	r.Delete("/api/v1/namespaces/{namespace}/limitranges/{name}", h.DeleteLimitRange)

	// MutatingWebhookConfiguration (cluster-scoped)
	r.Get("/api/v1/mutatingwebhookconfigurations", h.GetMutatingWebhookConfigurations)
	r.Get("/api/v1/mutatingwebhookconfigurations/{name}/describe", h.DescribeMutatingWebhookConfiguration)
	r.Get("/api/v1/mutatingwebhookconfigurations/{name}/yaml", h.GetMutatingWebhookConfigurationYAML)
	r.Delete("/api/v1/mutatingwebhookconfigurations/{name}", h.DeleteMutatingWebhookConfiguration)

	// ValidatingWebhookConfiguration (cluster-scoped)
	r.Get("/api/v1/validatingwebhookconfigurations", h.GetValidatingWebhookConfigurations)
	r.Get("/api/v1/validatingwebhookconfigurations/{name}/describe", h.DescribeValidatingWebhookConfiguration)
	r.Get("/api/v1/validatingwebhookconfigurations/{name}/yaml", h.GetValidatingWebhookConfigurationYAML)
	r.Delete("/api/v1/validatingwebhookconfigurations/{name}", h.DeleteValidatingWebhookConfiguration)

	// Lease (namespace-scoped)
	r.Get("/api/v1/leases/all", h.GetAllLeases)
	r.Get("/api/v1/namespaces/{namespace}/leases", h.GetLeases)
	r.Get("/api/v1/namespaces/{namespace}/leases/{name}/describe", h.DescribeLease)
	r.Get("/api/v1/namespaces/{namespace}/leases/{name}/yaml", h.GetLeaseYAML)
	r.Delete("/api/v1/namespaces/{namespace}/leases/{name}", h.DeleteLease)
}

File: frontend/src/services/api/configuration.ts
// Configuration API — cluster-wide config-ish resources that don't
// fit into workloads / network / storage / security:
// PriorityClass / RuntimeClass / Lease / ResourceQuota / LimitRange /
// MutatingWebhookConfiguration / ValidatingWebhookConfiguration /
// ConfigMap / Secret.

import { client } from './client'
import type {
  ConfigMapInfo,
  LeaseInfo,
  LimitRangeInfo,
  PriorityClassInfo,
  ResourceQuotaInfo,
  RuntimeClassInfo,
  SecretInfo,
  WebhookConfigInfo,
} from './types'

export const configurationApi = {
  // PriorityClass
  getPriorityClasses: async (forceRefresh = false): Promise<PriorityClassInfo[]> => {
    const { data } = await client.get('/cluster/priorityclasses', {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  describePriorityClass: async (name: string): Promise<any> => {
    const { data } = await client.get(`/cluster/priorityclasses/${name}/describe`)
    return data
  },

  deletePriorityClass: async (name: string): Promise<void> => {
    await client.delete(`/cluster/priorityclasses/${name}`)
  },

  // RuntimeClass
  getRuntimeClasses: async (forceRefresh = false): Promise<RuntimeClassInfo[]> => {
    const { data } = await client.get('/cluster/runtimeclasses', {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  describeRuntimeClass: async (name: string): Promise<any> => {
    const { data } = await client.get(`/cluster/runtimeclasses/${name}/describe`)
    return data
  },

  deleteRuntimeClass: async (name: string): Promise<void> => {
    await client.delete(`/cluster/runtimeclasses/${name}`)
  },

  // Lease
  getAllLeases: async (forceRefresh = false): Promise<LeaseInfo[]> => {
    const { data } = await client.get('/cluster/leases/all', {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  getLeases: async (namespace: string, forceRefresh = false): Promise<LeaseInfo[]> => {
    const { data } = await client.get(`/cluster/namespaces/${namespace}/leases`, {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  describeLease: async (namespace: string, name: string): Promise<any> => {
    const { data } = await client.get(`/cluster/namespaces/${namespace}/leases/${name}/describe`)
    return data
  },

  deleteLease: async (namespace: string, name: string): Promise<void> => {
    await client.delete(`/cluster/namespaces/${namespace}/leases/${name}`)
  },

  // ResourceQuotas
  getAllResourceQuotas: async (forceRefresh = false): Promise<ResourceQuotaInfo[]> => {
    const { data } = await client.get('/cluster/resourcequotas/all', {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  getResourceQuotas: async (namespace: string, forceRefresh = false): Promise<ResourceQuotaInfo[]> => {
    const { data } = await client.get(`/cluster/namespaces/${namespace}/resourcequotas`, {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  describeResourceQuota: async (namespace: string, name: string): Promise<any> => {
    const { data } = await client.get(`/cluster/namespaces/${namespace}/resourcequotas/${name}/describe`)
    return data
  },

  deleteResourceQuota: async (namespace: string, name: string): Promise<void> => {
    await client.delete(`/cluster/namespaces/${namespace}/resourcequotas/${name}`)
  },

  // LimitRanges
  getAllLimitRanges: async (forceRefresh = false): Promise<LimitRangeInfo[]> => {
    const { data } = await client.get('/cluster/limitranges/all', {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  getLimitRanges: async (namespace: string, forceRefresh = false): Promise<LimitRangeInfo[]> => {
    const { data } = await client.get(`/cluster/namespaces/${namespace}/limitranges`, {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  describeLimitRange: async (namespace: string, name: string): Promise<any> => {
    const { data } = await client.get(`/cluster/namespaces/${namespace}/limitranges/${name}/describe`)
    return data
  },

  deleteLimitRange: async (namespace: string, name: string): Promise<void> => {
    await client.delete(`/cluster/namespaces/${namespace}/limitranges/${name}`)
  },

  // MutatingWebhookConfigurations
  getMutatingWebhookConfigurations: async (forceRefresh = false): Promise<WebhookConfigInfo[]> => {
    const { data } = await client.get('/cluster/mutatingwebhookconfigurations', {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  describeMutatingWebhookConfiguration: async (name: string): Promise<any> => {
    const { data } = await client.get(`/cluster/mutatingwebhookconfigurations/${name}/describe`)
    return data
  },

  deleteMutatingWebhookConfiguration: async (name: string): Promise<void> => {
    await client.delete(`/cluster/mutatingwebhookconfigurations/${name}`)
  },

  // ValidatingWebhookConfigurations
  getValidatingWebhookConfigurations: async (forceRefresh = false): Promise<WebhookConfigInfo[]> => {
    const { data } = await client.get('/cluster/validatingwebhookconfigurations', {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  describeValidatingWebhookConfiguration: async (name: string): Promise<any> => {
    const { data } = await client.get(`/cluster/validatingwebhookconfigurations/${name}/describe`)
    return data
  },

  deleteValidatingWebhookConfiguration: async (name: string): Promise<void> => {
    await client.delete(`/cluster/validatingwebhookconfigurations/${name}`)
  },

  // ConfigMaps
  getConfigMaps: async (namespace: string, forceRefresh = false): Promise<ConfigMapInfo[]> => {
    const { data } = await client.get(`/cluster/namespaces/${namespace}/configmaps`, {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  getAllConfigMaps: async (forceRefresh = false): Promise<ConfigMapInfo[]> => {
    const { data } = await client.get('/cluster/configmaps/all', {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  describeConfigMap: async (namespace: string, name: string): Promise<any> => {
    const { data } = await client.get(`/cluster/namespaces/${namespace}/configmaps/${name}/describe`)
    return data
  },

  deleteConfigMap: async (namespace: string, name: string): Promise<void> => {
    await client.delete(`/cluster/namespaces/${namespace}/configmaps/${name}`)
  },

  // Secrets
  getSecrets: async (namespace: string, forceRefresh = false): Promise<SecretInfo[]> => {
    const { data } = await client.get(`/cluster/namespaces/${namespace}/secrets`, {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  getAllSecrets: async (forceRefresh = false): Promise<SecretInfo[]> => {
    const { data } = await client.get('/cluster/secrets/all', {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  describeSecret: async (namespace: string, name: string): Promise<any> => {
    const { data } = await client.get(`/cluster/namespaces/${namespace}/secrets/${name}/describe`)
    return data
  },

  getSecretYaml: async (namespace: string, name: string): Promise<{ yaml: string }> => {
    const { data } = await client.get(`/cluster/namespaces/${namespace}/secrets/${name}/yaml`)
    return typeof data === 'string' ? { yaml: data } : data
  },

  deleteSecret: async (namespace: string, name: string): Promise<void> => {
    await client.delete(`/cluster/namespaces/${namespace}/secrets/${name}`)
  },
}
